[v6r22] Using Proxy Providers and OAuthManager(OAuthDIRAC extension)

TaykYoku opened the pull request:
BEGINRELEASENOTES

*Configuration
CHANGE: add new helper methods for IdP/Proxy Providers, OAuthManager. Add methods that read DN options in dirac user section

*Framework
CHANGE: add possibility to work with Proxy Providers and "clean" proxy without voms/dirac extensions

ENDRELEASENOTES

fstagni commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -74,6 +84,16 @@ def __initializeDB(self):
                                      }
 
+    if 'ProxyDB_CleanProxies' not in tablesInDB:
+      tablesD['ProxyDB_CleanProxies'] = {'Fields': {'UserName': 'VARCHAR(64) NOT NULL',
```
This table would be for which kind of proxies?

TaykYoku replied:
This table need to store long proxies without extensions(like as dirac_group, voms)

fstagni commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -74,6 +84,16 @@ def __initializeDB(self):
+    if 'ProxyDB_CleanProxies' not in tablesInDB:
+      tablesD['ProxyDB_CleanProxies'] = {'Fields': {'UserName': 'VARCHAR(64) NOT NULL',
+                                                    'UserDN': 'VARCHAR(255) NOT NULL',
+                                                    'ProxyProvider': 'VARCHAR(64) NOT NULL',
```
Don't you need also a user group?

TaykYoku replied:
No, in case when the proxy is taken from this table the group is added later

fstagni commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -3,21 +3,24 @@
 
 __RCSID__ = "$Id$"
 
+import os
```
Changes to this DB should be validated through an integration test that should sit in /tests/Integration/Framework (to be created)

fstagni commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -596,13 +645,143 @@ def __getPUSProxy(self, userDN, userGroup, requiredLifetime, requestedVOMSAttr=N
 
+  def __getDIRACCAProxy(self, fullName, eMail, DN=None, diracGroup=None, proxyProvider='DIRAC_EOSC_CA'):
+
+    def writeUserConfigFile(fileName, fullName, eMail):
```
This method deserves an extended explanation

TaykYoku replied:
It method was deleted. Instead, it uses ProxyProviderFactory().getProxyProvider(proxyProvider)

TaykYoku commented:
Depends on #4070

TaykYoku commented on ConfigurationSystem/Client/Helpers/Resources.py:
```diff
@@ -370,3 +370,43 @@ def getFilterConfig(filterID):
+
+def getProxyProviders():
+  """ Get the names of all the Proxy Providers defined in the CS """
+  return gConfig.getSections('/Resources/ProxyProviders')
```
done

fstagni commented:
The rst file should be linked from this one: https://github.com/DIRACGrid/DIRAC/blob/rel-v6r22/docs/source/AdministratorGuide/Configuration/ConfReference/Resources/index.rst

Also, fix the unit test that fails.

fstagni commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -395,28 +430,34 @@ def storeProxy(self, userDN, userGroup, chain):
+      dValues['UserGroup'] = sUserGroup
+      dValues['PersistentFlag'] = "'False'"
+    else:
+      dValues['ProxyProvider'] = "'%s'" % proxyProvider
```
`dValues['PersistentFlag']` shouldn't be set in this case?

TaykYoku replied:
In this case table store proxy generated from scratch. I think, it do not allow MyProxy to be used in the near future for cleanProxies table

fstagni commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -596,18 +672,86 @@ def __getPUSProxy(self, userDN, userGroup, requiredLifetime, requestedVOMSAttr=N
+
+        :return: S_OK(dict)/S_ERROR() -- dict with remaining secudnds, proxy as string and as chain
+    """
+    gLogger.notice('Getting proxy for "%s" DN by "%s" proxy provider' % (userDN, proxyProvider))
```
```suggestion
    gLogger.info('Getting proxy from proxyProvider', '(for "%s" DN by "%s")' % (userDN, proxyProvider))
```

TaykYoku replied:
done

chaen commented:
I've already told @atsareg last week, but it would be nice if you could run all the proxy tests that are in v7 against this new type of proxy, and adapt the tests accordingly

fstagni commented:
Can you point at exactly which tests? And it would be great if you can identify the commit(s) that introduced them and cherry-picking in rel-v6r22 branch.

chaen commented:
Not so trivial because they are testing both M2Crypto and pygsi, so cannot really be backported cleanly. The tests are here: https://github.com/DIRACGrid/DIRAC/tree/integration/Core/Security/test

fstagni commented:
I see, and it seems to me that only a subset of those can be used here. @TaykYoku would you like to have a look at what applies here? We are very careful with changes to proxies.

TaykYoku commented:
> The rst file should be linked from this one: https://github.com/DIRACGrid/DIRAC/blob/rel-v6r22/docs/source/AdministratorGuide/Configuration/ConfReference/Resources/index.rst

done

> Also, fix the unit test that fails.

I cannot do it without #4070 PR integration, because I need import ProxyProvider resources:
 "ImportError: No module named ProxyProvider.ProxyProviderFactory"

atsareg commented:
Can we defer the integration tests to v7r0 release just to go quicker with v6r22 with OAuthDIRAC as extension ? In v7r0 I hope we will be able to merge it into the core DIRAC and then include integration tests.

chaen commented on tests/Integration/Framework/Test_ProxyDB.py:
```diff
@@ -0,0 +1,179 @@
+from DIRAC.FrameworkSystem.DB.ProxyDB import ProxyDB
+
+
+diracTestCACFG = """
```
Could you reuse the same test proxies and everything that is already in integration branch ? It would be much easier. Once ported to integration, we can just point to the existing location

TaykYoku replied:
The v7 integration test/certs has been added to use it in an existing test, I have not found any ProxyDB tests or that can be use for ProxyDB testing. Could you give me test files links from v7 that needed to integrate to this PR or what thing additionally need to test

chaen replied:
They are not tests for the proxyDB but for the X509 objects that carries your PEM identity. It is all the `Test_X509*` files in https://github.com/DIRACGrid/DIRAC/tree/integration/Core/Security/test 
It would be interesting that you run these tests against a *clean proxy* that you get from a provider. 
Of course, these tests were designed to compare and run against PyGSI and M2Crypto for the existing cases, so probably you need to extend them

atsareg replied:
Tests for the DIRAC CA are included into #4070. This is the main use case. Testing RCAuth proxies will anyway require mocking-up the RCAuth proxy and will be reduced to the similar use case as DIRAC CA proxies

TaykYoku commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -623,6 +767,11 @@ def getProxy(self, userDN, userGroup, requiredLifeTime=False):
 
     # Standard proxy is requested
     retVal = self.__getPemAndTimeLeft(userDN, userGroup)
+    if retVal['OK'] and requiredLifeTime:
```
here I try to describe this logic:
    ####
    # if not get proxy from DB
    if not retVal['OK']:
      # try to generate it with requiredLifeTime
      retVal = self.__generateProxy(userDN, userGroup, requiredLifeTime=requiredLifeTime)
      if not retVal['OK']:
        # if no lets crash
        return retVal
    # if we get proxy from DB(in ganerate case it not need)
    elif requiredLifeTime:
      # and have requiredLifeTime
      if retVal['Value'][1] < requiredLifeTime and not self.__useMyProxy:
        # and it more than we have in proxy from DB we try to generate new
        retVal = self.__generateProxy(userDN, userGroup, requiredLifeTime=requiredLifeTime)
        if not retVal['OK']:
          # if no lets crash
          return retVal
    # all ok, continue 
    ####

TaykYoku commented on ConfigurationSystem/Client/Helpers/Registry.py:
```diff
@@ -379,7 +379,13 @@ def getVOMSRoleGroupMapping(vo=''):
 
 
-def getUsernameForID(ID, usersList=False):
+def getUsernameForID(ID, usersList=[]):
```
ok

TaykYoku commented on ConfigurationSystem/Client/Helpers/Registry.py:
```diff
@@ -396,3 +402,54 @@ def getCAForUsername(username):
+  """
+  result = getUsernameForDN(DN)
+  if not result['OK']:
+    return []
```
ok

TaykYoku commented on ConfigurationSystem/Client/Helpers/Registry.py:
```diff
@@ -396,3 +402,54 @@ def getCAForUsername(username):
+  if not result['OK']:
+    return []
+  # Convert DN to section format
+  secDN = DN.replace('/', '-').replace('=', '_')
```
done

TaykYoku commented on ConfigurationSystem/Client/Helpers/Registry.py:
```diff
@@ -396,3 +402,54 @@ def getCAForUsername(username):
+  """
+  result = getUsernameForDN(DN)
+  if not result['OK']:
+    return []
```
done

TaykYoku commented on ConfigurationSystem/Client/Helpers/Resources.py:
```diff
@@ -370,3 +370,54 @@ def getFilterConfig(filterID):
   return gConfig.getOptionsDict('Resources/LogFilters/%s' % filterID)
+
+
+def getProxyProviders(providerName=None):
```
deleted

TaykYoku commented on ConfigurationSystem/Client/Helpers/Resources.py:
```diff
@@ -370,3 +370,54 @@ def getFilterConfig(filterID):
+  return gConfig.getSections('%s/ProxyProviders' % gBaseResourcesSection)
+
+
+def getProxyProviderOption(proxyProvider, Value):
```
done

TaykYoku commented on ConfigurationSystem/Client/Helpers/Resources.py:
```diff
@@ -370,3 +370,54 @@ def getFilterConfig(filterID):
+  return gConfig.getValue("%s/ProxyProviders/%s/%s" % (gBaseResourcesSection, proxyProvider, Value))
+
+
+def getProxyProviderDict(IdP):
```
done

TaykYoku commented on ConfigurationSystem/Client/Helpers/Resources.py:
```diff
@@ -370,3 +370,54 @@ def getFilterConfig(filterID):
+  return gConfig.getValue("%s/IdProviders/%s/%s" % (gBaseResourcesSection, providerName, option))
+
+
+def getIdPSections(IdP, path=''):
```
done

TaykYoku commented on ConfigurationSystem/Client/Helpers/Resources.py:
```diff
@@ -370,3 +370,54 @@ def getFilterConfig(filterID):
+  return gConfig.getSections("%s/IdProviders/%s/%s" % (gBaseResourcesSection, IdP, path))
+
+
+def getIdPOptions(IdP, path=''):
```
done

TaykYoku commented on ConfigurationSystem/Client/Helpers/Resources.py:
```diff
@@ -370,3 +370,54 @@ def getFilterConfig(filterID):
+  return gConfig.getOptions("%s/IdProviders/%s/%s" % (gBaseResourcesSection, IdP, path))
+
+
+def getIdPDict(IdP):
```
done

TaykYoku commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -438,48 +479,83 @@ def purgeExpiredProxies(self, sendNotifications=True):
-  def deleteProxy(self, userDN, userGroup='any'):
+  def deleteProxy(self, userDN, userGroup='any', proxyProvider=None):
     """ Remove proxy of the given user from the repository
+
```
Now it`ll be use both arguments

TaykYoku commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -438,48 +479,83 @@ def purgeExpiredProxies(self, sendNotifications=True):
+    if proxyProvider:
+      cmd += ' AND ProxyProvider="%s"' % proxyProvider
+    else:
+      if userGroup and not proxyProvider:
```
you're right

chaen commented:
@atsareg for changes in the proxy DB, I really do not think we can go without integration tests. These changes impact each and every installation of DIRAC. Moreover, you know that if tests and docs don't make it at the same time as the code, they never do :-) 
Frankly speaking, the correct thing to do would have been to hurry v7r0 out before pushing all that, because you would have at least a minimum test coverage already.

TaykYoku commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -438,48 +479,83 @@ def purgeExpiredProxies(self, sendNotifications=True):
+          result = chain.generateProxyToString(record[1], diracGroup=userGroup, rfc=True)
+          if not result['OK']:
+            return result
+          chain.loadProxyFromString(result['Value'])
```
It remained from the debugging, fixed now

TaykYoku commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -438,48 +479,83 @@ def purgeExpiredProxies(self, sendNotifications=True):
+          if not result['OK']:
+            return result
+          chain.loadProxyFromString(result['Value'])
+          return S_OK((result['Value'], record[1]))
```
Proxy provider provide proxy only to store it in DB. Other logic is original with except one case, when not found proxy in default DBs it`ll look in cleanDB and add ext. e.g. group

atsareg replied:
@chaen : we want to optimize the proxy storage with respect to two aspects: minimizing the number of requests to proxy providers and minimizing the number of different proxies to store. That is why we cache proxies coming from proxy providers and refresh them only if need would be. And second, we store proxies without the DIRAC group embedded and will add group according to the particular request to the ProxyManager and after the check for the group membership validity. This in fact should simplify the logic of the proxy storage and provisioning.

andresailer commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -623,6 +773,11 @@ def getProxy(self, userDN, userGroup, requiredLifeTime=False):
+      if retVal['Value'][1] < requiredLifeTime and not self.__useMyProxy:
+        retVal = self.__generateProxy(userDN, userGroup, requiredLifeTime=requiredLifeTime)
+    else:
+      retVal = self.__generateProxy(userDN, userGroup, requiredLifeTime=requiredLifeTime)
```
So, `if retVal['Value'][1] >= requiredLifeTime` (the if above is not entered), but I don't have proxyProviders, this will return an error, and thus this whole function will return S_ERROR?

TaykYoku replied:
In this case retVal['OK'] is true, so block (if not retVal['OK']: return retVal) will be ignore

andresailer replied:
Right, I messed up which `if` this `else` belongs to, the problem actually happens when `requiredLifeTime == False`, then even though `retVal['OK']` is true one ends up in this `else`, or am I again missing something?

TaykYoku replied:
Code in last redaction:
# retVal['OK'] == True & requiredLifeTime == False

if retVal['OK'] and requiredLifeTime:
  # in this case we ignore it
  if retVal['Value'][1] < requiredLifeTime and not self.__useMyProxy:
    errMsg = 'the required lifetime is less than the time left in the proxy'
    retVal = self.__generateProxyFromProxyProviders(userDN, userGroup, requiredLifeTime=requiredLifeTime)

else:
  # and come here
  errMsg = retVal['Message']
  # where we try to generate proxy with requiredLifeTime=False
  retVal = self.__generateProxyFromProxyProviders(userDN, userGroup, requiredLifeTime=requiredLifeTime)

if retVal['OK']:
  # all ok, continue

I do not see the problem, maybe I do not understand what you mean

andresailer replied:
https://github.com/TaykYoku/DIRAC/blob/185a34d41d645d7bcb4ed6fd815b0ce2e8531e96/FrameworkSystem/DB/ProxyDB.py#L766-L773

```python
    requiredLifeTime  = False # From the function argument
    retVal = self.__getPemAndTimeLeft(userDN, userGroup)  # returns S_OK
    if retVal['OK'] and requiredLifeTime: # is not entered
      if retVal['Value'][1] < requiredLifeTime and not self.__useMyProxy:
        errMsg = 'the required lifetime is less than the time left in the proxy'
        retVal = self.__generateProxyFromProxyProviders(userDN, userGroup, requiredLifeTime=requiredLifeTime)
    else: # we end up here
      errMsg = retVal['Message'] # exception KeyError (retVal is S_OK(), doesn't have 'Message', but this is a different problem now)
      retVal = self.__generateProxyFromProxyProviders(userDN, userGroup, requiredLifeTime=requiredLifeTime) 
      # if exception before doesn't happen, this now returns S_ERROR, because
      # I don't have ProxyProviders defined, retVal is now S_ERROR instead of S_OK
```
One has to assume that there are cases out there where this function was previously called with requiredLifeTime evaluating to False, which would not result in an error (otherwise we can just return S_ERROR at the top...)

TaykYoku replied:
In this case, we don`t have a proxy in the DB and cannot use MyProxy and cannot generate a new one, so the method just ended in an error that should catch the last line:
https://github.com/TaykYoku/DIRAC/blob/185a34d41d645d7bcb4ed6fd815b0ce2e8531e96/FrameworkSystem/DB/ProxyDB.py#L789-L790
```
    return S_ERROR("Can't get a proxy%s %s" % (requiredLifeTime and ' for %s seconds:' % requiredLifeTime or ':', errMsg))
```

TaykYoku replied:
o, I seem to understand what you mean

TaykYoku replied:
This logic will be correct

```
    retVal = self.__getPemAndTimeLeft(userDN, userGroup)
    if not retVal['OK']:
      errMsg = retVal['Message']
      retVal = self.__generateProxyFromProxyProviders(userDN, userGroup, requiredLifeTime=requiredLifeTime)
    elif requiredLifeTime:
      if retVal['Value'][1] < requiredLifeTime and not self.__useMyProxy:
        errMsg = 'the required lifetime is less than the time left in the proxy'
        retVal = self.__generateProxyFromProxyProviders(userDN, userGroup, requiredLifeTime=requiredLifeTime)
    if retVal['OK']:
```

andresailer replied:
Yes, this seems to not turn S_OK to S_ERROR. I will have another look once you pushed your changes.

TaykYoku commented on ConfigurationSystem/Client/Helpers/Registry.py:
```diff
@@ -396,3 +402,54 @@ def getCAForUsername(username):
+  for DN in result['Value']:
+    if proxyProvider in getProxyProvidersForDN(DN):
+      return S_OK(DN)
+  return S_ERROR("No DN found for %s proxy provider for user ID %s" % (proxyProvider, userID))
```
done

TaykYoku commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -342,18 +365,21 @@ def storeProxy(self, userDN, userGroup, chain):
+      proxyGroup = retVal['Value']
+      if not proxyGroup:
+        proxyGroup = Registry.getDefaultUserGroup()
+      if not userGroup == proxyGroup:
```
done

TaykYoku commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -596,18 +677,87 @@ def __getPUSProxy(self, userDN, userGroup, requiredLifetime, requestedVOMSAttr=N
+    PPList = result['Value']
+    for proxyProvider in PPList:
+      result = self.__getPemAndTimeLeft(userDN, userGroup, proxyProvider=proxyProvider)
+      if result['OK']:
```
thanks, it really looks better

TaykYoku commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -596,18 +677,87 @@ def __getPUSProxy(self, userDN, userGroup, requiredLifetime, requestedVOMSAttr=N
+        if not result['OK']:
+          return result
+        return S_OK((result['Value'], remainingSecs))
+    return result
```
fixed

TaykYoku commented on ConfigurationSystem/Client/Helpers/Resources.py:
```diff
@@ -370,3 +370,83 @@ def getFilterConfig(filterID):
+def getProxyProviders():
+  """ Return list of all proxy provider names defined in the CS
+
+      :return: S_OK(lisl)/S_ERROR
```
fixed

TaykYoku commented on ConfigurationSystem/Client/Helpers/Registry.py:
```diff
@@ -379,7 +379,14 @@ def getVOMSRoleGroupMapping(vo=''):
+def getUsernameForID(ID, usersList=None):
+  """ Get DIRAC user name by ID
+
+      :param basestring DN: user DN
```
done

TaykYoku commented on ConfigurationSystem/Client/Helpers/Registry.py:
```diff
@@ -392,7 +399,84 @@ def getUsernameForID(ID, usersList=False):
+
+
+def getDNSectionName(userDN):
+  """ Change user DN string by replacing spécial symbol that not used in
```
done

TaykYoku commented on ConfigurationSystem/Client/Helpers/Registry.py:
```diff
@@ -392,7 +399,84 @@ def getUsernameForID(ID, usersList=False):
+  if not result['Value']:
+    return S_ERROR('No proxy providers found for "%s" user DN' % userDN)
+  ppList = result['Value']
+  if not isinstance(result['Value'], list):
```
done

TaykYoku commented on ConfigurationSystem/Client/Helpers/Registry.py:
```diff
@@ -392,7 +399,84 @@ def getUsernameForID(ID, usersList=False):
+
+
+def getDNProperty(userDN, value):
+  """ Get property from PNProperties section by user DN
```
done

TaykYoku commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -596,18 +677,87 @@ def __getPUSProxy(self, userDN, userGroup, requiredLifetime, requestedVOMSAttr=N
+        chain = result['Value']['chain']
+        remainingSecs = result['Value']['remainingSecs']
+        result = chain.generateProxyToString(remainingSecs, diracGroup=userGroup, rfc=True)
+        if not result['OK']:
```
fixed

TaykYoku commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -596,18 +677,87 @@ def __getPUSProxy(self, userDN, userGroup, requiredLifetime, requestedVOMSAttr=N
+      return S_OK({'proxy': proxyStr, 'chain': chain, 'remainingSecs': remainingSecs})
+    return result
+
+  def __generateProxy(self, userDN, userGroup, requiredLifeTime):
```
done

TaykYoku commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -623,6 +773,11 @@ def getProxy(self, userDN, userGroup, requiredLifeTime=False):
     retVal = self.__getPemAndTimeLeft(userDN, userGroup)
+    if retVal['OK'] and requiredLifeTime:
+      if retVal['Value'][1] < requiredLifeTime and not self.__useMyProxy:
+        retVal = self.__generateProxy(userDN, userGroup, requiredLifeTime=requiredLifeTime)
```
fixed

TaykYoku commented on docs/source/AdministratorGuide/Configuration/ConfReference/Resources/ProxyProviders/index.rst:
```diff
@@ -0,0 +1,62 @@
+|                         | framework(https://tools.ietf.org/html/rfc6749)    |                                                                               |
++-------------------------+---------------------------------------------------+-------------------------------------------------------------------------------+
+| <specific parameter>    | Some specific parameter for specific proxy        | max_proxylifetime = 864000                                                    |
+|                         | provider installation, e.g. CheckIns MasterPortal | proxy_endpoint = https://masterportal-pilot.aai.egi.eu/mp-oa2-server/getproxy |                                                           |
```
done

TaykYoku commented on docs/source/AdministratorGuide/Configuration/ConfReference/Resources/ProxyProviders/index.rst:
```diff
@@ -0,0 +1,62 @@
++-------------------------+---------------------------------------------------+-------------------------------------------------------------------------------+
+| **Name**                | **Description**                                   | **Example**                                                                   |
++-------------------------+---------------------------------------------------+-------------------------------------------------------------------------------+
+| issuer                  | Authorization server's issuer identifier URL	    | https://masterportal-pilot.aai.egi.eu/mp-oa2-server                           |
```
done

atsareg commented:
ProxyDB integration tests, If I understand correctly, should be written from scratch. I agree, that this should be done, of course. But we need a v6r22 prerelease with these PRs merged also to develop thise integration tests. So, I suggest that we merge these PRs now and will use the resulting prerelease for various testing that eventually will be incorporated as integration tests.

chaen commented:
`But we need a v6r22 prerelease with these PRs merged also to develop thise integration tests.` Why ? Unless I am missing something, you can develop your integration tests without having the code released

atsareg commented:
The real tests are done with real proxy providers, and these tests show what should be actually tested. In the integration tests just some mock-up cases can be implemented which is a subset of real tests. To me obviously we should start with real tests. Just to be pragmatic and go quicker.

chaen commented:
Look, I already gave my opinion about all that, and I did not change my mind: this work is fantastically useful and I am really glad that it is done. At the same time, it impacts every VO in a dangerous way, and that's why it was proposed you create an extension (even for the proxyDB, you can do it easily), such that you can do quick prototyping.
It apparently was decided to keep the ProxyDB changes in vanilla DIRAC. So either we all obey the rules (docs + tests + certification et tutti quanti) or we make exceptions (like that one). As I said, I gave my opinion already, but at this point, it is not up to me (after all, the technical coordinator was re elected unanimously for another 2 years :-) ). 

I think we all want to be pragmatic and go quicker (yet some PR wait for months or even years, and so do some eagerly waited releases, despite being optional features, with environment variables and whatnot to protect the existing behavior). 

I do not want to start a flame-throwing discussion on all that. Again, I really appreciate the work, but I do feel concerned (and, to be bluntly honest, annoyed to postpone yet again v7, but that's another unrelated topic). 
I'm done with my review and giving my opinion :-) 

PS: I'll be very interested though in having feedback once you actually use that ! :-)

andresailer commented:
I strongly propose to write unit tests for the getProxy function (and maybe others were new things are injected) that covers all branches, and ensures backward compatibility for configurations that do not contain ProxyProviders. This at least can show isolation of the additions from the existing functionality.

fstagni commented:
At the workshop I showed how to use Jenkins for running integration tests on non-released code. This has been used for few years for several developments by several users and work perfectly. So, I don't think there's an issue there.

andresailer commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -624,22 +765,30 @@ def getProxy(self, userDN, userGroup, requiredLifeTime=False):
+              errMsg = retVal['Message']
+            else:
+              errMsg = "the required lifetime is less than the time left in the proxy"
+    return S_ERROR("Can't get a proxy%s %s" % (requiredLifeTime and ' for %s seconds:' % requiredLifeTime or ':',
```
This indentation is off. It would be very useful to have unit tests that will spot this kind if issue (apologies if you are already working on them.)

TaykYoku replied:
fixed and tested

andresailer commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -624,22 +765,30 @@ def getProxy(self, userDN, userGroup, requiredLifeTime=False):
+      if retVal['Value'][1] < requiredLifeTime and not self.__useMyProxy:
+        errMsg = 'the required lifetime is less than the time left in the proxy'
+        retVal = self.__generateProxyFromProxyProviders(userDN, userGroup, requiredLifeTime=requiredLifeTime)
+    if retVal['OK']:
```
If you do
```
if not retVal['OK']:
  return retVal
```
You can save on one indentation level in the next block

TaykYoku replied:
fixed

andresailer commented on FrameworkSystem/DB/ProxyDB.py:
```diff
@@ -624,22 +765,30 @@ def getProxy(self, userDN, userGroup, requiredLifeTime=False):
+      timeLeft = retVal['Value'][1]
+      chain = X509Chain()
+      retVal = chain.loadProxyFromString(pemData)
+      if retVal['OK']:
```
similarly here, return on error, keep going otherwise, save on indentation

TaykYoku replied:
fixed

fstagni commented on tests/Integration/Framework/Test_ProxyManager.py:
```diff
@@ -0,0 +1,340 @@
+    print('Try to upload proxy with next params:\n Proxy: %s\nRFC: %s\nLivetime: %s\nUploadParams:\n%s\nProxyParams:\n%s\n' %
+          (proxy, rfcReq, timeReq, uploadParams, proxyParams))
+    result = self.proxyMgr.uploadProxy(proxy=proxy, restrictLifeTime=timeReq, rfcIfPossible=rfcReq)
+    print('\nRESULT: %s' % result)
```
No assert(s) here?

TaykYoku replied:
the tests have been fixed and implemented in Test_ProxyDB.py

fstagni commented on tests/Integration/Framework/Test_ProxyManager.cfg:
```diff
@@ -0,0 +1,123 @@
+Resources
```
Don't call this file "Test_*" (or pytest will try to execute it)

TaykYoku replied:
removed

fstagni commented on tests/Integration/Framework/Test_ProxyManager.cfg:
```diff
@@ -0,0 +1,123 @@
+    DIRAC_TEST_CA
+    {
+      ProxyProviderType = DIRACCA
+      CertFile = ../../../Core/Security/test/certs/ca/ca.cert.pem
```
I assume here you are looking into all those files that you added in DIRAC/Core/Security/test. If those files are used only by integration tests please move them somewhere under DIRAC/tests/Integration.

What is in DIRAC/Core/Security/test should be unit test only (basically what is run in Travis) with pytest.

TaykYoku replied:
Certs moved to Integration section

fstagni commented on tests/Integration/Framework/Test_ProxyDB.py:
```diff
@@ -0,0 +1,280 @@
+                                                   dbParam['Host'],
+                                                   dbParam['Port'],
+                                                   dbParam['DBName']))
+connection = engine.connect()
```
I don't get why you are using sqlalchemy here.

TaykYoku replied:
not use now

fstagni commented on FrameworkSystem/Client/ProxyManagerClient.py:
```diff
@@ -147,43 +153,34 @@ def uploadProxy(self, proxy=False, diracGroup=False, chainToConnect=False, restr
-    return S_OK()
+    return S_OK(result.get('proxies') or result['Value'])
 
+  # FIXME: Old method
```
If there are old methods that are unused within DIRAC (BTW did you check WebAppDIRAC?) then don't put a #FIXME, rather use the `deprecated` decorator (search the code for it)

TaykYoku replied:
It's my work comment that I forgot to remove.

TaykYoku replied:
Yes, I remember about WebAppDIRAC. I`ll do it.

fstagni commented on FrameworkSystem/scripts/dirac-proxy-get-uploaded-info.py:
```diff
@@ -33,8 +33,10 @@ def setUser(arg):
+userName = userName or proxyProps.get('username')
 if not userName:
-  userName = proxyProps['username']
+  print "Your proxy don`t have username extension"
```
no `print`s please.

TaykYoku replied:
fixed

fstagni commented on FrameworkSystem/scripts/dirac-proxy-init.py:
```diff
@@ -36,12 +36,17 @@ def setUploadPilotProxy( self, _arg ):
     Script.registerSwitch( "U", "upload", "Upload a long lived proxy to the ProxyManager", self.setUploadProxy )
     Script.registerSwitch( "P", "uploadPilot", "Upload a long lived pilot proxy to the ProxyManager", self.setUploadPilotProxy )
     Script.registerSwitch( "M", "VOMS", "Add voms extension", self.setVOMSExt )
+    #Script.registerSwitch( "g:", "group=", "DIRAC Group to embed in the proxy", self.setDIRACGroup )
```
What? This is widely used.

TaykYoku replied:
"g, group=" keys still present and use in dirac-proxy-init

fstagni commented on tests/System/client_proxy.sh:
```diff
@@ -0,0 +1,71 @@
+
+# This is a generic list of commands run from a client for testing proxy manager client.
+#
+# Submitter should follow through the logs
```
There's also something similar in https://github.com/DIRACGrid/DIRAC/blob/rel-v6r22/tests/System/client_core.sh so in case you have more tests just modify that file, which is what is run in the DIRAC certification.

TaykYoku replied:
done

TaykYoku commented on tests/Integration/Framework/Test_ProxyManager.sh:
```diff
@@ -0,0 +1,25 @@
+#!/bin/sh
```
Testing from shell moved to Systems/client_core.sh

TaykYoku commented:
> Did you remove the integration test of the ProxyDB in the last commits?
Now integration tests present in the Test_ProxyDB.py and client_core.sh

TaykYoku commented:
Since "[v6r22] ProxyProvider resources #4070" PR will not evolve over the next week, it was decided to merge it with the available to complete the integration of the changes faster
Also it's help to complete the Travis CI build checks

fstagni commented on tests/System/client_core.sh:
```diff
@@ -4,106 +4,156 @@
+echo "========================"
+echo "===  dirac-proxy-init -U"
+echo
+dirac-proxy-init -C $SERVERINSTALLDIR/user/client.pem -K $SERVERINSTALLDIR/user/client.key -U
```
This is a system test, and normally $SERVERINSTALLDIR is not defined, as normally it's run within the certification process, defined in this trello card: https://trello.com/c/1IPtPhds (the environment there is simply what results in a DIRAC client installation, so created with dirac-install and nothing else).

If you want to run this in Jenkins, then you should use not $SERVERINSTALLDIR but $TESTCODE.

TaykYoku replied:
Test has been changed to run on Jenkins and if variables will not defined

fstagni commented:
Hi, since there have been few changes in the machinery for running tests on Jenkins, I updated the test 
https://jenkins-dirac.web.cern.ch/view/Development/job/Andrei_proxy_providers/
and tried to run it (please check its configuration)

It fails because dirac-proxy-init tries to upload a proxy before the ProxyManager is even installed (https://jenkins-dirac.web.cern.ch/view/Development/job/Andrei_proxy_providers/15/console). I am not sure I understand why it's needed (is it a new default?).
If I ignore it and keep going the integration test for ProxyDB passes: https://jenkins-dirac.web.cern.ch/view/Development/job/Andrei_proxy_providers/ws/

TaykYoku commented:
> Hi, since there have been few changes in the machinery for running tests on Jenkins, I updated the test
> https://jenkins-dirac.web.cern.ch/view/Development/job/Andrei_proxy_providers/
> and tried to run it (please check its configuration)
> 
> It fails because dirac-proxy-init tries to upload a proxy before the ProxyManager is even installed (https://jenkins-dirac.web.cern.ch/view/Development/job/Andrei_proxy_providers/15/console). I am not sure I understand why it's needed (is it a new default?).
> If I ignore it and keep going the integration test for ProxyDB passes: https://jenkins-dirac.web.cern.ch/view/Development/job/Andrei_proxy_providers/ws/

"Andrei_proxy_providers" was created for understand how to use Jenkins.
Working space where I try to run Jenkins tests is "Andrii_modifyProxyServiceToUseOAuth2_v6r22".

fstagni commented:
Alright, so please compare what's in https://jenkins-dirac.web.cern.ch/view/Development/job/Andrei_proxy_providers/configure with https://jenkins-dirac.web.cern.ch/view/Development/job/Andrii_modifyProxyServiceToUseOAuth2_v6r22/configure and update the few things that need to be updated. 

At the same time it seems to me that  "Andrii_modifyProxyServiceToUseOAuth2_v6r22" is broken. And I think that what I showed above is anyway valid.

fstagni commented:
Please rebase on `integration` branch.

TaykYoku commented:
> Alright, so please compare what's in https://jenkins-dirac.web.cern.ch/view/Development/job/Andrei_proxy_providers/configure with https://jenkins-dirac.web.cern.ch/view/Development/job/Andrii_modifyProxyServiceToUseOAuth2_v6r22/configure and update the few things that need to be updated.
> 
> At the same time it seems to me that "Andrii_modifyProxyServiceToUseOAuth2_v6r22" is broken. And I think that what I showed above is anyway valid.

Now Jenkins test is working

fstagni commented on tests/Integration/Framework/Test_ProxyDB.py:
```diff
@@ -0,0 +1,543 @@
+
+    gLogger.info('* Upload proxy..')
+    for user, dn, group, vo, time, res, log in [("user_1", '/C=DN/O=DIRAC/CN=user_1', "group_1", False, 12,
+                                                 False, 'With group extansion'),
```
`extension` not `extansion` (here and later too)

TaykYoku replied:
done

fstagni commented on tests/Integration/Resources/ProxyProvider/Test_DIRACCAProxyProvider.py:
```diff
@@ -0,0 +1,192 @@
+
+if __name__ == '__main__':
+  SUITE = unittest.defaultTestLoader.loadTestsFromTestCase(DIRACCAPPTest)
+  unittest.TextTestRunner(verbosity=2).run(SUITE)
```
Last line should be:
```
sys.exit(not testResult.wasSuccessful())
```

TaykYoku replied:
done

fstagni commented:
OK, I see. I have a question: why do you have these lines in the test:
``
cp -r $TESTCODE/DIRAC/ResourceStatusSystem/Utilities $SERVERINSTALLDIR/DIRAC/ResourceStatusSystem/
cp -r $TESTCODE/DIRAC/Interfaces/API/DiracAdmin.py  $SERVERINSTALLDIR/DIRAC/Interfaces/API/
cp -r $TESTCODE/DIRAC/ConfigurationSystem/Client $SERVERINSTALLDIR/DIRAC/ConfigurationSystem/
cp -r $TESTCODE/DIRAC/Resources/ProxyProvider  $SERVERINSTALLDIR/DIRAC/Resources/
cp -r $TESTCODE/DIRAC/FrameworkSystem $SERVERINSTALLDIR/DIRAC/
cp -r $TESTCODE/DIRAC/Core/Utilities $SERVERINSTALLDIR/DIRAC/Core/
```
?

I don't really they are needed. 
BTW: DIRAC_ALTERNATIVE_SRC_ZIP is deprecated, that's why I was asking you to use the other Jenkins job, where I had updated with the correct configuration.


Anyway, I know it's a painful back-and-forth, but can you rebase on integration? v6r22 is out now, and this PR won't be merged in the rel-v6r22 branch. Next week we'll start certification of v7r0, and this won't take to long to get into a production release.

TaykYoku commented:
> OK, I see. I have a question: why do you have these lines in the test:
> 
> ```
> cp -r $TESTCODE/DIRAC/ResourceStatusSystem/Utilities $SERVERINSTALLDIR/DIRAC/ResourceStatusSystem/
> cp -r $TESTCODE/DIRAC/Interfaces/API/DiracAdmin.py  $SERVERINSTALLDIR/DIRAC/Interfaces/API/
> cp -r $TESTCODE/DIRAC/ConfigurationSystem/Client $SERVERINSTALLDIR/DIRAC/ConfigurationSystem/
> cp -r $TESTCODE/DIRAC/Resources/ProxyProvider  $SERVERINSTALLDIR/DIRAC/Resources/
> cp -r $TESTCODE/DIRAC/FrameworkSystem $SERVERINSTALLDIR/DIRAC/
> cp -r $TESTCODE/DIRAC/Core/Utilities $SERVERINSTALLDIR/DIRAC/Core/
> ```
> 
> ?
> 
> I don't really they are needed.
>
It's for replacing original code of running services(DB also) to changed code from my branch. After restart of services tests can testing new methods and fitches.

> BTW: DIRAC_ALTERNATIVE_SRC_ZIP is deprecated, that's why I was asking you to use the other Jenkins job, where I had updated with the correct configuration.
>
Fixed now, I somehow missed that moment

> Anyway, I know it's a painful back-and-forth, but can you rebase on integration? v6r22 is out now, and this PR won't be merged in the rel-v6r22 branch. Next week we'll start certification of v7r0, and this won't take to long to get into a production release.
>
No pain) Yes, I understand that this PR will no longer be integrated with v6r22, but is it possible for v6r23?
Rebase with integration return too many conflicts may be better to make new PR?

fstagni commented:
We'll not make v6r23: we'll go straight to v7r0. There's a lot of code in there developed long time ago by several developers, that really needs to go in production. We have done pretty well with certifying v6r22, so I hope we'll do the same with v7r0. 

The purpose of DIRAC_ALTERNATIVE_SRC_ZIP is to replace the code for running services, DBs, and so on, so basically copying your code (which you did - again - by hand). At the same time now dirac-install.py can do the same thing, when called with "--module" (or "-m") flag. In jenkins you can use "ALTERNATIVE_MODULES" for invoking that flag (have a look at https://github.com/DIRACGrid/DIRAC/blob/integration/Core/scripts/dirac-install.py#L92).
Also, I see that you used "fullInstallDIRAC". If you use "miniInstallDIRAC" it's faster.
